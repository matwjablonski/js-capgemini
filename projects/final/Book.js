class BookAbstract {
  data;

  constructor(data) {
    this.data = data;

    if (new.target === BookAbstract) {
      throw new Error("Nie można utworzyć instancji klasy Book");
    }

    this.prepareBookItem();
  }

  prepareBookItem() {
    throw new Error("Should be implemented in child class");
  }

  handleFavorite() {
    throw new Error("Should be implemented in child class");
  }
}

class Book extends BookAbstract {
  container;
  data;

  constructor(data) {
    super(data);

    this.data = data;
    this.prepareBookItem();
  }

  prepareBookItem() {
    const bookItem = document.createElement("li");
    bookItem.classList.add("book-item");

    if (this.data.isFavorite) {
      bookItem.classList.add("favorite");
    }

    this.container = bookItem;

    this.renderTitle();
    this.renderAuthor();
    this.renderCategory();

    this.renderActionButton();
  }

  renderTitle() {
    const bookTitle = document.createElement("h2");
    bookTitle.dataset.testId = "test-id-book-title";
    bookTitle.textContent = this.data.title;

    this.container.appendChild(bookTitle);
  }

  renderAuthor() {
    const {
      data: {
        author,
      },
      container,
    } = this;

    const bookAuthor = document.createElement("p");
    bookAuthor.textContent = author;

    container.appendChild(bookAuthor);
  }
  
  renderCategory() {
    const bookCategory = document.createElement("p");
    bookCategory.textContent = `Kategoria: ${this.data.category}`;
    bookCategory.classList.add("book-category");
  
    this.container.appendChild(bookCategory);
  }

  renderActionButton() {
    const action = document.createElement("button");
    action.textContent = this.prepareTextForActionButton();
  
    action.classList.add("action-button");
    this.container.appendChild(action);

    action.addEventListener('click', this.handleFavorite.bind(this));
  }

  prepareTextForActionButton() {
    return this.data.isFavorite ? "Usuń z ulubionych" : "Dodaj do ulubionych";
  }

  handleFavorite(event) {
    const {
      container,
    } = this;
    this.data.isFavorite = !this.data.isFavorite;
    event.target.textContent = this.prepareTextForActionButton();
    container.classList.toggle("favorite");
  }
}

export default Book;
