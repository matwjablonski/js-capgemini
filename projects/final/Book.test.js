import Book from './Book';

describe('Book', () => {
  let container;
  let book;

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    const data = { title: 'test', author: 'test', category: 'test', isFavorite: false };
    book = new Book(data);
    container.appendChild(book.container);
  });

  afterEach(() => {
    document.body.removeChild(container);
  });

  test('should create a book item', () => {
    expect(book.container.classList.contains('book-item')).toBe(true);
  });

  test('should render book title', () => {
    const title = book.container.querySelector('[data-test-id="test-id-book-title"]');
    expect(title.textContent).toBe(book.data.title);
  });
});
