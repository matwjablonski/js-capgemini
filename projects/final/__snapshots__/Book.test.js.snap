// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book should match the snapshot 1`] = `
<li
  class="book-item"
>
  <h2
    data-test-id="test-id-book-title"
  >
    test
  </h2>
  <p
    data-test-id="test-id-book-author"
  >
    test
  </p>
  <p
    class="book-category"
  >
    Kategoria: test
  </p>
  <button
    class="action-button"
  >
    Dodaj do ulubionych
  </button>
</li>
`;
